modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbMvPartitionStorage.java

```diff
@@ -0,0 +1,494 @@
+    /** {@inheritDoc} */
+    @Override
+    public Cursor<BinaryRow> scan(Predicate<BinaryRow> keyFilter, @Nullable Timestamp timestamp) throws StorageException {
+        // Set next partition as an upper bound.
+        var options = new ReadOptions().setIterateUpperBound(upperBound).setTotalOrderSeek(true);
+
+        RocksIterator it = db.newIterator(cf, options);
+
+        // Seek iterator to the beginning of the partition.
+        it.seek(partitionStartPrefix());
+
+        // Size of the buffer to seek values. Should fit partition id, row id and maybe timestamp, if it's not null.
+        int seekKeyBufSize = ROW_ID_OFFSET + igniteRowIdSize + (timestamp == null ? 0 : TIMESTAMP_SIZE);
+
+        // Here's seek buffer itself. Originally it contains a valid partition id, row id payload that's filled with zeroes, and maybe
+        // a timestamp value. Zero row id guarantees that it's lexicographically less then or equal to any other row id stored in the
+        // partition.
+        // Byte buffer from a thread-local field can't be used here, because of two reasons:
+        //  - no one guarantees that there will only be a single cursor;
+        //  - no one guarantees that returned cursor will not be used other threads.
+        // The thing is, we need this buffer to preserve its content between invocactions of "hasNext" method.
+        ByteBuffer seekKeyBuf = ByteBuffer.allocate(seekKeyBufSize).order(BIG_ENDIAN).putShort((short) partId);
+
+        if (timestamp != null) {
+            putTimestamp(seekKeyBuf.position(ROW_ID_OFFSET + igniteRowIdSize), timestamp);
+        }
+
+        // Version without timestamp to compare row ids.
+        ByteBuffer seekKeyBufSliceWithoutTimestamp = timestamp == null
+                ? seekKeyBuf
+                : seekKeyBuf.position(0).limit(ROW_ID_OFFSET + igniteRowIdSize).slice().order(BIG_ENDIAN);
+
+        return new Cursor<BinaryRow>() {
+            /** Cached value for {@link #next()} method. Also optimizes the code of {@link #hasNext()}. */
+            private BinaryRow next = null;
+
+            /** {@inheritDoc} */
+            @Override
+            public boolean hasNext() {
+                // Fastpath for consecutive invocations.
+                if (next != null) {
+                    return true;
+                }
+
+                // Prepare direct buffer slice to read keys from the iterator. It's goind to be used to read row ids without timestamps.
+                // To enforce this, we limit it to a proper capacity.
+                ByteBuffer directBuffer = MV_KEY_BUFFER.get()
+                        .position(0).limit(ROW_ID_OFFSET + igniteRowIdSize).slice().order(BIG_ENDIAN);
+
+                while (true) {
+                    // This flag is used to skip row ids that were created after required timestamp.
+                    boolean found = true;
+
+                    // We should do it all the time. Here in particular it means one of two things:
```

Comment:
'all the time'? Wouldn't 'after each seek' be clearer?

Reply:
Maybe